libsolutil/LP.cpp

```diff
@@ -0,0 +1,1613 @@
+
+		if (!removeEmptyColumns(_state, _model, changed))
+			return false;
+	}
+
+	// TODO return the values selected for named variables in order to
+	// be used when returning the model.
+	return true;
+}
+
+void normalizeRowLengths(SolvingState& _state)
+{
+	size_t vars = max(_state.variableNames.size(), _state.bounds.size());
+	for (Constraint const& c: _state.constraints)
+		vars = max(vars, c.data.size());
+	_state.variableNames.resize(vars);
+	_state.bounds.resize(vars);
+	for (Constraint& c: _state.constraints)
+		c.data.resize(vars);
+}
+
+
+}
+
+bool Constraint::operator<(Constraint const& _other) const
+{
+	if (equality != _other.equality)
+		return equality < _other.equality;
+
+	for (size_t i = 0; i < max(data.size(), _other.data.size()); ++i)
+	{
+		 rational const& a = get(data, i);
+		 rational const& b = get(_other.data, i);
+		 if (a != b)
+			return a < b;
+	}
+	return false;
+}
+
+bool Constraint::operator==(Constraint const& _other) const
+{
+	if (equality != _other.equality)
+		return false;
+
+	for (size_t i = 0; i < max(data.size(), _other.data.size()); ++i)
+		 if (get(data, i) != get(_other.data, i))
+			return false;
+	return true;
+}
+
+bool SolvingState::operator<(SolvingState const& _other) const
+{
+	if (variableNames == _other.variableNames)
+	{
+		if (bounds == _other.bounds)
+			return constraints < _other.constraints;
+		else
+			return bounds < _other.bounds;
+	}
+	else
+		return variableNames < _other.variableNames;
+}
+
+bool SolvingState::operator==(SolvingState const& _other) const
+{
+	return
+		variableNames == _other.variableNames &&
+		bounds == _other.bounds &&
+		constraints == _other.constraints;
+}
+
+string SolvingState::toString() const
+{
+	string result;
+
+	for (Constraint const& constraint: constraints)
+	{
+		vector<string> line;
+		for (auto&& [index, multiplier]: constraint.data | ranges::views::enumerate)
+			if (index > 0 && multiplier != 0)
+			{
+				string mult =
+					multiplier == -1 ?
+					"-" :
+					multiplier == 1 ?
+					"" :
+					::toString(multiplier) + " ";
+				line.emplace_back(mult + variableNames.at(index));
+			}
+		result += joinHumanReadable(line, " + ") + (constraint.equality ? "  = " : " <= ") + ::toString(constraint.data.front()) + "\n";
+	}
+	result += "Bounds:\n";
+	for (auto&& [index, bounds]: bounds | ranges::view::enumerate)
+	{
+		if (!bounds[0] && !bounds[1])
+			continue;
+		if (bounds[0])
+			result += ::toString(*bounds[0]) + " <= ";
+		result += variableNames.at(index);
+		if (bounds[1])
+			result += " <= " + ::toString(*bounds[1]);
+		result += "\n";
+	}
+	return result;
+}
+
+
+pair<bool, map<size_t, bool>> DPLL::simplify()
+{
+	// TODO use vector?
+	map<size_t, bool> assignments;
+	while (!clauses.empty() && !ranges::all_of(clauses, [](Clause const& _c) { return _c.literals.size() > 1; }))
+	{
+		// TODO this assumes that no clause contains more than one constraint
+		// TODO add an assertion for that.
+
+		for (Clause const& c: clauses)
```

Comment:
Is this loop doing unit propagation or handling pure literals?

Reply:
I assume you meant `true` / `false` by "pure literals"? The data structure currently cannot model them.

Reply:
By pure I mean literals that have the same polarity in all clauses

Reply:
No, it only handles empty clauses or clauses that consist only of a single literal. Literals with the same polarity sound like an interesting optimization opportunity!